File: Leetcode/2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
	public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
		ListNode newNode, prevNode = null, retNode = null;
		boolean extra1 = false;
		boolean first = true;
		while (l1 != null && l2 != null) {
			newNode = new ListNode(l1.val + l2.val);
			if (first) {
				retNode = newNode;
				first = false;
			}
			if (extra1) {
				newNode.val++;
				extra1 = false;
			}
			if (newNode.val >= 10) {
				extra1 = true;
				newNode.val -= 10;
			}
			if (prevNode != null) prevNode.next = newNode;
			prevNode = newNode;
			l1 = l1.next;
			l2 = l2.next;
		}
		while (l1 != null) {
			newNode = new ListNode(l1.val);
			if (extra1) {
				newNode.val ++;
				extra1 = false;
			}
			if (newNode.val >= 10) {
				extra1 = true;
				newNode.val -= 10;
			}
			prevNode.next = newNode;
			prevNode = newNode;
			l1 = l1.next;
		}
		while (l2 != null) {
			newNode = new ListNode(l2.val);
			if (extra1) {
				newNode.val ++;
				extra1 = false;
			}
			if (newNode.val >= 10) {
				extra1 = true;
				newNode.val -= 10;
			}
			prevNode.next = newNode;
			prevNode = newNode;
			l2 = l2.next;
		}
		if (extra1) {
			newNode = new ListNode(1);
			prevNode.next = newNode;
			prevNode = newNode;
		}
		return retNode;
	}
}

File: Leetcode/2.cc
#include<bits/stdc++.h>
#pragma GCC optimize ("Ofast")

using namespace std;

static int fast_io = [] () {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 0;
} ();

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* head = new ListNode(-1);
        ListNode* it = head;
        bool carry = false;
        while (l1 || l2) {
            
            ListNode* new_node = new ListNode((l1 ? l1 -> val : 0) + (l2 ? l2 -> val : 0) + (carry ? 1 : 0));
            carry = false;
            it -> next = new_node; it = it -> next;
            
            if (new_node -> val >= 10) {
                carry = true;
                new_node -> val -= 10;
            }
            
            if (l1) l1 = l1 -> next;
            if (l2) l2 = l2 -> next;
            
        }
        
        if (carry) it -> next = new ListNode(1);
        
        return head -> next;
        
    }
};
